레스토랑 조회 구현하기

dongakim opened the pull request:
**UI**

- [x] 지역 리스트
- [x] 카테고리 리스트
- [x] 레스토랑 리스트

**Logics**

- [x] 지역 조회
- [x] 카테고리 조회
- [x] 지역 선택
- [x] 카테고리 선택
- [x] 레스토랑 조회

dongakim committed:
RestaurantRegionsContainer 테스트 추가

dongakim committed:
RestaurantRegionsContainer 구현

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

hannut91 commented:
TDD를 하실 때 리팩터링을 꼭 잊지말고 수행하셔야 합니다 ㅎㅎ

dongakim committed:
RestaurantRegionsContainer 테스트 추가

dongakim committed:
react-redux, redux-thunk 적용

dongakim committed:
api 구현

dongakim committed:
actions 구현

dongakim committed:
react-redux mock 구현

dongakim committed:
RestaurantRegionsContainer 구현

dongakim committed:
RestaurantRegionsContainer 적용

dongakim committed:
actions 리팩토링

dongakim committed:
RestaurantsCategoriesContainer 테스트 추가

dongakim committed:
RestaurantsCategoriesContainer 구현

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

dongakim committed:
reducer 리팩토링

action type 상수화

dongakim committed:
RestaurantsCategoriesContainer 테스트 추가

dongakim committed:
api 추가

dongakim committed:
async action 추가

dongakim committed:
RestaurantsCategoriesContainer 구현

dongakim committed:
RestaurantsCategoriesContainer 적용

dongakim committed:
api 리팩토링

dongakim committed:
RestaurantsContainer 테스트 추가

dongakim committed:
RestaurantsContainer 구현

dongakim committed:
RestaurantRegionsContainer 테스트 추가

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

dongakim committed:
RestaurantRegionsContainer 구현

dongakim committed:
action 리팩토링

action type, action creator 생성

dongakim committed:
RestaurantRegionsContainer 테스트 추가

dongakim committed:
reducer 수정

dongakim committed:
RestaurantRegionsContainer 구현

dongakim committed:
RestaurantRegionsContainer 리팩토링

dongakim committed:
RestaurantsCategoriesContainer 테스트 추가

dongakim committed:
RestaurantsCategoriesContainer 구현

dongakim committed:
RestaurantsCategoriesContainer 테스트 추가

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

dongakim committed:
RestaurantsCategoriesContainer 구현

dongakim committed:
리팩토링

action type, action creator 생성

dongakim committed:
RestaurantsCategoriesContainer 리팩토링

dongakim committed:
RestaurantsContainer 테스트 추가

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

dongakim committed:
RestaurantsContainer 구현

dongakim committed:
RestaurantsContainer 적용

dongakim committed:
리팩토링

action type 생성

dongakim committed:
RestaurantsCategoriesContainer 테스트 추가

dongakim committed:
api 추가

dongakim committed:
async action 추가

dongakim committed:
RestaurantsCategoriesContainer 구현

dongakim committed:
RestaurantRegionsContainer 테스트 수정

dongakim committed:
RestaurantRegionsContainer 구현

dongakim committed:
actions 수정

dongakim committed:
reducer 테스트 추가

dongakim committed:
reducer 구현

dongakim commented:
<img width="682" alt="스크린샷 2020-10-20 오전 9 49 24" src="https://user-images.githubusercontent.com/22358415/96526813-9b60e380-12b9-11eb-8be8-f884b946fdef.png">
api 테스트, action 테스트는 어떻게 진행해야하는지 궁금합니다.

hannut91 commented:
액션에서 주요 관심사가 무엇인지 생각해보시면 좋을 것 같습니다. 예를들어서 레스토랑 목록을 가져오는 것을 볼게요.

```js
export const loadRestaurants = () => async (dispatch, getState) => {
  const restaurants = await restaurantService.getRestaurants();
  
  dispatch(setRestaurants(restaurants));
}
```

이 비동기 액션에서 하는 일은
* 서버에서 레스토랑 목록을 가져온다.
* 가져온 레스토랑 목록을 setRestaurants액션을 통해 dispatch한다.

restaurantService에서 레스토랑 목록을 요청한 후 레스토랑을 업데이트 하는 setRestauarnts라는 액션을 다시 dispatch합니다. 하지만 레스토랑 목록을 어떻게 가져오는지는 주요한 관심사가 아닙니다. 그래서 getRestaurants는 mocking하여 처리하고 올바르게 가져왔다고 가정하고 액션이 올바르게 처리되는지 테스트합니다.

```js
import thunk from 'redux-thunk';

import configureStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const restaurants = [
  { id: 1, name: '마녀주방', category: '한식', address: '강남' },
];

getRestaurants.mockResvoledValue(restaurants);

test('loadRestaurants', async () => {
  store = mockStore({});
  
  await loadRestaurants();
  
  const actions = store.getActions();
  
  expect(actions[0]).toEqual(setRestaurants(restaurants));
});
```

See also
  *  https://github.com/reduxjs/redux-mock-store

hannut91 commented:
`redux-mock-store`를 사용하지 않고 `dispatch`를 mocking해서 할 수 있지만 ㅎㅎ 안 쓸 이유가 없습니다.

dongakim committed:
redux-mock-store 추가

dongakim committed:
api mock 구현

dongakim committed:
actions 테스트 추가

dongakim committed:
actions 리팩토링

dongakim commented:
services의 api는 보통 어떻게 테스트하나요?

hannut91 commented on src/RestaurantRegionsContainer.test.jsx:
```diff
@@ -0,0 +1,62 @@
+
+import RestaurantRegionsContainer from './RestaurantRegionsContainer';
+
+beforeEach(() => {
+  jest.clearAllMocks();
+});
```
이 로직은 `describe`안에 있는게 더 자연스러울 것 같아요.

hannut91 commented on src/RestaurantRegionsContainer.test.jsx:
```diff
@@ -0,0 +1,62 @@
+
+describe('<RestaurantRegionsContainer />', () => {
+  const dispatch = jest.fn();
+  useDispatch.mockImplementation(() => dispatch);
+
+  useSelector.mockImplementation((selector) => selector({
+    regions: [{ id: 1, name: '서울' }],
+    selectedRegion: { id: 1, name: '서울' },
+  }));
```
각 테스트마다 동일한 환경에서 테스트가 되도록 이 로직들도 `beforeEach`안에 넣어주는게 좋을 것 같아요

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,54 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === SET_REGIONS) {
```
`if`나 `switch`없이 객체를 이용해서 구현해보는건 어떨까요?

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,131 @@
+      const previousState = {
+        regions: [],
+      };
+      const action = {
+        type: SET_REGIONS,
+        payload: [{ id: 1, name: '서울' }],
+      };
```
액션 크리에이터를 이용해서 작성하는건 어떨까요?

hannut91 commented on src/services/api.js:
```diff
@@ -0,0 +1,17 @@
+}
+
+export function fetchRestaurantRegion() {
+  return fetcher('https://eatgo-customer-api.ahastudio.com/regions');
```
`fetcher`라는 어떤 동작이라기 보다는 동작을 하는 주체인 것 처럼 보여요. 마치 `fetcher.fetch`인 것 같은 느낌이 들어요.

dongakim committed:
테스트 리팩토링

beforeEach를 더 자연스럽게 describe 안으로 이동

dongakim committed:
테스트 리팩토링

각 테스트마다 동일한 환경에서 테스트가 되도록 변경

dongakim committed:
reducer 테스트 리팩토링

dongakim committed:
reducer 리팩토링

if, switch 문을 사용하지 않고 reducer 구현
참고: https://read.reduxbook.com/markdown/part1/03-updating-state.html

dongakim committed:
api 테스트 추가

dongakim committed:
api 리팩토링

의미에 맞게 함수명 변경

hannut91 commented on src/RestaurantRegionsContainer.test.jsx:
```diff
@@ -0,0 +1,63 @@
+    });
+
+    // Then
+    expect(dispatch).toBeCalledTimes((getAllByRole('button').length) * 2 + 1);
```
2 * 1이 이해하기 어려운 것 같아요. redux mock store를 이용해서 mocking하면 어떤 액션들이 호출되었는지 테스트할 수 있긴 합니다

dongakim replied:
이 부분은 redux mock store을 활요해서 구현해보려했는데 어렵네요..

hannut91 commented on src/RestaurantsContainer.jsx:
```diff
@@ -0,0 +1,17 @@
+
+  return (
+    <ul>
+      {restaurants.map((restaurant) => (
```
구조 분해 할당으로 반복되는 restauarnt를 없애는 건 어떨까요?

hannut91 commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,31 @@
+
+describe('<RestaurantsContainer />', () => {
+  useSelector.mockImplementation((selector) => selector({
+    restaurants: [{
```
반복적으로 사용되는 테스트용 데이터들은 다른 곳에 미리 선언해두고 사용하면 좋을 것 같아오

hannut91 commented on src/actions.js:
```diff
@@ -0,0 +1,63 @@
+};
+
+export const loadRestaurantCategories = () => async (dispatch) => {
+  const data = await fetchRestaurantCategories();
```
카테고리 목록을 가져오니 categories는 어떨까요?

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,116 @@
+        setRestaurantCategory({ id: 1, name: '한식' }),
+      );
+
+      expect(state.selectedCategory).toMatchObject({ id: 1, name: '한식' });
```
toEqual matcher를 사용하는 건 어떨까요?

dongakim replied:
이거는 특별한 이유가 있을까요?

hannut91 replied:
용도가 다르고 또 의도가 다를 것 같아요. 지금은 완전히 같은지 비교하는게 더 맞을 것 같아요. 

See also
  - https://jestjs.io/docs/en/expect#tomatchobjectobject

hannut91 reviewed (changes requested):
테스트용 데이터를 fixtrues라는 폴더로 분리하면 복잡성도 줄어들고 재사용성도 좋아질것 같아요 ㅎㅎ

dongakim committed:
리팩토링

restaurant 구조분해할당

dongakim committed:
리팩토링

데이터 fixture 생성

dongakim committed:
리팩토링

API 응답값 데이터명 시멘틱하게 변경

hannut91 commented on src/actions.test.js:
```diff
@@ -0,0 +1,100 @@
+        // Then
+        const actions = store.getActions();
+
+        expect(actions[0]).toEqual(undefined);
```
`actions`가 비어있다고 테스트하는건 어떨까요?

`expect(actions).toHaveLength(0);`

hannut91 reviewed (approved):
과제 풀이 영상을 통해서 복습해보시면 좋을 것 같습니다 ㅎㅎ 

이번 과제에서는 무엇을 배우셨나요? 다음에 더 시도해보고 싶은 것은 무엇인가요?
